extensions/amp-analytics/0.1/cookie-writer.js

```diff
@@ -0,0 +1,136 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {isObject, toWin} from '../../../src/types';
+import {hasOwn, map} from '../../../src/utils/object';
+import {user} from '../../../src/log';
+import {Services} from '../../../src/services';
+import {setCookie} from '../../../src/cookies';
+import {BASE_CID_MAX_AGE_MILLIS} from '../../../src/service/cid-impl';
+import {isInFie} from '../../../src/friendly-iframe-embed';
+import {Deferred} from '../../../src/utils/promise';
+import {isProxyOrigin} from '../../../src/url';
+
+
+const TAG = 'amp-analytics/cookie-writer';
+
+const EXPAND_WHITELIST = {
+  'QUERY_PARAM': true,
+  // TODO: Add linker_param
+}
+
+export class CookieWriter {
+
+  /**
+   * @param {!Window} win
+   * @param {!Element} element
+   * @param {!JsonObject} config
+   */
+  constructor (win, element, config) {
+    /** @private {!Window} */
+    this.win_ = win;
+
+    /** @private {!Element} */
+    this.element_ = element;
+
+    /** @private {!../../../src/service/url-replacements-impl.UrlReplacements} */
+    this.urlReplacementService_ = Services.urlReplacementsForDoc(element);
+
+    /** @private {!Array<!Promise>} */
+    this.promises_ = [];
+
+    /** @private {!../../../src/utils/promise.Deferred} */
+    this.readyPromise_ = new Deferred();
+
+    this.init_(config);
```

Comment:
instead of doing all the work in constructor, we can move it to `whenReady`. Actually, we'd rename `whenReady` to `write` or `run`

```
run() {
  ...
  return readyPromise;
}

```